// vite-plugin-vue-transition-root-validator/dist/index.cjs
let node_url = require("node:url");
let node_fs = require("node:fs");

//#region src/i18n.ts
const translations = {
	zh: {
		messageHeader: "[vite-plugin-vue-transition-root-validator] 检测到 Vue Transition 多根节点错误",
		setupInstructions: `
如何在项目中启用此插件：

1. 在 vite.config.ts 中添加插件：
   import vitePluginVueTransitionRootValidator from 'vite-plugin-vue-transition-root-validator';

   plugins: [
     vitePluginVueTransitionRootValidator()
   ]

2. 在 src/main.ts 中初始化：
   import { setupVueRootValidator } from 'virtual:vue-transition-root-validator';

   const app = createApp(App);
   setupVueRootValidator(app, { lang: 'zh' });
   app.mount('#app');
`
	},
	en: {
		messageHeader: "[vite-plugin-vue-transition-root-validator] Vue Transition Multiple Root Nodes Error",
		setupInstructions: `
How to enable this plugin in your project:

1. Add plugin in vite.config.ts:
   import vitePluginVueTransitionRootValidator from 'vite-plugin-vue-transition-root-validator';

   plugins: [
     vitePluginVueTransitionRootValidator()
   ]

2. Initialize in src/main.ts:
   import { setupVueRootValidator } from 'virtual:vue-transition-root-validator';

   const app = createApp(App);
   setupVueRootValidator(app, { lang: 'en' });
   app.mount('#app');
`
	}
};
function getMessageHeader(lang) {
	return translations[lang].messageHeader;
}

//#endregion
//#region src/index.ts
/**
* 发送错误覆盖层到客户端（优先定向发送，避免首屏时广播命中不到当前 client）
*/
function sendErrorOverlay(args) {
	const { server, message, lang, client } = args;
	const payload = {
		type: "error",
		err: {
			message: getMessageHeader(lang),
			stack: message
		}
	};
	if (client?.send) {
		client.send(payload);
		return;
	}
	server.ws.send(payload);
}
/**
* Vite 插件：Vue Root Validator
*
* 用于检测 Vue 组件在 <Transition> 内渲染时的多根节点问题
*
* @returns Vite 插件对象
*
* @example
* ```ts
* // vite.config.ts
* import vitePluginVueTransitionRootValidator from 'vite-plugin-vue-transition-root-validator';
*
* export default defineConfig({
*   plugins: [
*     vitePluginVueTransitionRootValidator()
*   ]
* });
* ```
*/
function vitePluginVueTransitionRootValidator() {
	let resolved;
	return {
		name: "vite-plugin-vue-transition-root-validator",
		apply: "serve",
		configResolved(config) {
			resolved = config;
		},
		configureServer(server) {
			if (resolved.command !== "serve") return;
			const lastByClient = /* @__PURE__ */ new WeakMap();
			server.ws.on("vite-plugin-vue-transition-root-validator:vue-warn", (payload, client) => {
				if (!payload?.message) return;
				const key = payload.key ?? payload.message.slice(0, 800);
				if (lastByClient.get(client) === key) {
					client?.send?.("vite-plugin-vue-transition-root-validator:ack", { key });
					return;
				}
				lastByClient.set(client, key);
				const effectiveLang = payload.lang ?? "en";
				sendErrorOverlay({
					server,
					message: payload.message,
					lang: effectiveLang,
					client
				});
				client?.send?.("vite-plugin-vue-transition-root-validator:ack", { key });
			});
		},
		resolveId(id) {
			if (id === "virtual:vue-transition-root-validator") return "\0virtual:vue-transition-root-validator";
			return null;
		},
		load(id) {
			if (id === "\0virtual:vue-transition-root-validator") {
				const clientEntryTs = (0, node_url.fileURLToPath)(new URL("./client.ts", require("url").pathToFileURL(__filename).href));
				const clientEntryJs = (0, node_url.fileURLToPath)(new URL("./client.js", require("url").pathToFileURL(__filename).href));
				const clientUrl = `/@fs/${((0, node_fs.existsSync)(clientEntryTs) ? clientEntryTs : clientEntryJs).replace(/\\/g, "/")}`;
				return `
// 虚拟模块：vue-transition-root-validator
// 此模块由 vite-plugin-vue-transition-root-validator 插件自动生成

import { setupVueRootValidator } from ${JSON.stringify(clientUrl)};

// 重新导出函数
export { setupVueRootValidator };
`;
			}
			return null;
		}
	};
}

//#endregion
module.exports = vitePluginVueTransitionRootValidator;
